src/redux/utils/connect-extension-middleware.ts

```diff
@@ -10,71 +9,102 @@ import { isEqual } from 'lodash';
-        if (!isEqual(preferences, lastSentState?.preferences)) {
-            statePatch.preferences = preferences;
-        }
-        if (!isEqual(contacts, lastSentState?.contacts)) {
-            statePatch.contacts = contacts;
+
+        const trimmedContacts = trimContacts(contacts);
+        if (!isEqual(trimmedContacts, lastSentState?.trimmedContacts)) {
+            statePatch.trimmedContacts = trimmedContacts;
         }
-        if (!isEqual(tokens, lastSentState?.tokens)) {
-            statePatch.tokens = tokens;
+
+        const trimmedPreferences = trimPreferences(preferences);
+        if (!isEqual(trimmedPreferences, lastSentState?.trimmedPreferences)) {
+            statePatch.trimmedPreferences = trimmedPreferences;
         }
 
-        if (!isEqual(tokens, lastSentState?.tokens)) {
-            statePatch.tokens = tokens;
+        const trimmedTokens = trimTokens(tokens);
+        if (!isEqual(trimmedTokens, lastSentState?.trimmedTokens)) {
+            statePatch.trimmedTokens = trimmedTokens;
         }
 
         if (statePatch) {
-            lastSentState = { ...lastSentState, ...statePatch };
+            lastSentState = {
+                ...lastSentState,
+                ...statePatch
+            };
         }
 
         return statePatch;
     }
 );
 
+let timer;
+const CONN_EXTENSION_STATE = `${CONN_EXTENSION}State`;
+
 export const connectExtensionMiddleware = store => next => async action => {
     next(action);
 
-    if (Platform.OS !== 'web' && (await ConnectExtensionWeb.isConnected())) {
-        const statePatch = getStatePatch(store.getState());
+    timer && clearTimeout(timer);
+
+    timer = setTimeout(async () => {
+        if (Platform.OS !== 'web' && (await ConnectExtensionWeb.isConnected())) {
+            const statePatch = getStatePatch(store.getState());
+
+            const currentStateHash = await hash(JSON.stringify(statePatch));
 
-        if (Object.keys(statePatch).length > 0) {
             try {
-                const connection = await ConnectExtensionWeb.getConnection();
+                const storageStateHash = await getItemFromStorage(CONN_EXTENSION_STATE);
 
-                if (connection) {
-                    await ConnectExtension.syncExtension(connection);
+                if (storageStateHash && storageStateHash !== currentStateHash) {
+                    // Sync Extension with the new state
+                    if (Object.keys(statePatch).length > 0) {
+                        try {
+                            const connection = await ConnectExtensionWeb.getConnection();
+
+                            if (connection) {
+                                await ConnectExtension.syncExtension(connection);
+                            }
+                        } catch {
+                            lastSentState = {};
+                        }
+                    }
+
+                    // Save State Hash to Async Storage
+                    // @ts-ignore - ignore for web
```

Comment:
why do we need the ignore ?

Reply:
Because of this strange ts error on ext:

![image](https://user-images.githubusercontent.com/32679726/81895007-d861d500-95b9-11ea-8534-15fb07d7c2f6.png)

`Argument of type 'WordArray' is not assignable to parameter of type 'string'.`

A solution would be to cast this as:
`await storeItemToStorage(currentStateHash as any, CONN_EXTENSION_STATE as any);`